packages/dd-trace/src/opentelemetry/metrics/periodic_metric_reader.js

```diff
@@ -0,0 +1,353 @@
+          unit,
+          type,
+          instrumentationScope,
+          temporality: this.#getTemporality(type),
+          data: [],
+          dataPointMap: new Map()
+        })
+      }
+
+      const metric = metricsMap.get(metricKey)
+
+      if (type === METRIC_TYPES.COUNTER || type === METRIC_TYPES.UPDOWNCOUNTER) {
+        this.#aggregateSum(metric, value, attributes, attrKey, timestamp, stateKey, cumulativeState)
+      } else if (type === METRIC_TYPES.HISTOGRAM) {
+        this.#aggregateHistogram(metric, value, attributes, attrKey, timestamp, stateKey, cumulativeState)
+      } else {
+        this.#aggregateLastValue(metric, value, attributes, attrKey, timestamp)
+      }
+    }
+
+    const metrics = [...metricsMap.values()]
+
+    this.#applyDeltaTemporality(metrics, lastExportedState)
+
+    return metrics
+  }
+
+  #getScopeKey (instrumentationScope) {
+    return `${instrumentationScope.name}@${instrumentationScope.version}@${instrumentationScope.schemaUrl}`
+  }
+
+  #getStateKey (scopeKey, name, type, attrKey) {
+    return `${scopeKey}:${name}:${type}:${attrKey}`
+  }
+
+  #isDeltaType (type) {
+    return type === METRIC_TYPES.COUNTER ||
+           type === METRIC_TYPES.OBSERVABLECOUNTER ||
+           type === METRIC_TYPES.HISTOGRAM
+  }
+
+  #applyDeltaTemporality (metrics, lastExportedState) {
+    for (const metric of metrics) {
+      if (metric.temporality === TEMPORALITY.DELTA && this.#isDeltaType(metric.type)) {
+        const scopeKey = this.#getScopeKey(metric.instrumentationScope)
+
+        for (const dataPoint of metric.data) {
+          const stateKey = this.#getStateKey(scopeKey, metric.name, metric.type, dataPoint.attrKey)
+
+          if (metric.type === METRIC_TYPES.COUNTER || metric.type === METRIC_TYPES.OBSERVABLECOUNTER) {
+            const lastValue = lastExportedState.get(stateKey) || 0
+            const currentValue = dataPoint.value
+            dataPoint.value = currentValue - lastValue
+            lastExportedState.set(stateKey, currentValue)
+          } else if (metric.type === METRIC_TYPES.HISTOGRAM) {
+            const lastState = lastExportedState.get(stateKey) || {
+              count: 0,
+              sum: 0,
+              bucketCounts: new Array(dataPoint.bucketCounts.length).fill(0)
+            }
+            const currentState = {
+              count: dataPoint.count,
+              sum: dataPoint.sum,
+              min: dataPoint.min,
+              max: dataPoint.max,
+              bucketCounts: [...dataPoint.bucketCounts]
+            }
+            dataPoint.count = currentState.count - lastState.count
+            dataPoint.sum = currentState.sum - lastState.sum
+            dataPoint.bucketCounts = currentState.bucketCounts.map(
+              (count, idx) => count - (lastState.bucketCounts[idx] || 0)
+            )
+            lastExportedState.set(stateKey, currentState)
+          }
+        }
+      }
+
+      delete metric.dataPointMap
```

Comment:
I am uncertain why that property is removed. Could you add a comment? I would also likely just set it to undefined anyway.

Reply:
This field was meant as a temporary store for datapoints to allow quick access.
metric.data is the list that gets encoded and sent in OTLP payloads, while metric.dataPointMap is a helper map for fast lookups within that list.

If a datapoint already exists in the list, we update it; if not, we create a new one.
The idea was to maintain both structures for performance reasons, but it ends up causing more confusion than benefit. I’ll revert this and just use dataPointMap, converting its values to a list right before encoding.